# HOMEWORK 3 to Lesson 6
## Задание к Уроку 6 " Обучение с учителем в Scikit-learn"

In [2]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from pylab import rcParams as rc

%matplotlib inline
%config InlineBackend.figure_format = 'svg'

### Задание 1

Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn.

- Создайте датафреймы X и Y из этих данных.
- Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
- Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
- Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
- Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [3]:
from sklearn.datasets import load_boston
boston = load_boston()

In [4]:
boston.keys() # просмотр содержимого датасета BOSTON HOUSE Prices

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [5]:
data = boston.data # Данные о недвижимости ()(значения различных наблюдений о них)
features = boston.feature_names # Признаки
target = boston.target # Целева япеременная (цена на недвижимость)

In [6]:
data.shape

(506, 13)

In [7]:
features

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [8]:
target

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [9]:
X = pd.DataFrame(data, columns = features) # Для оценик целевой перемнной выбираем значения признаков, созданных на основание данных data
Y = pd.DataFrame(target,columns = (['price']))

In [10]:
X

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [11]:
Y

price
0     24.0
1     21.6
2     34.7
3     33.4
4     36.2
..     ...
501   22.4
502   20.6
503   23.9
504   22.0
505   11.9

[506 rows x 1 columns]

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
# Разбиение выборки на тестовую и тренировчную
x_test,x_train,y_test,y_train = train_test_split(X, Y, test_size = 0.3, random_state = 42)

In [14]:
x_test

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
5     0.02985   0.0   2.18   0.0  0.458  6.430  58.7  6.0622   3.0  222.0   
116   0.13158   0.0  10.01   0.0  0.547  6.176  72.5  2.7301   6.0  432.0   
45    0.17142   0.0   6.91   0.0  0.448  5.682  33.8  5.1004   3.0  233.0   
16    1.05393   0.0   8.14   0.0  0.538  5.935  29.3  4.4986   4.0  307.0   
468  15.57570   0.0  18.10   0.0  0.580  5.926  71.0  2.9084  24.0  666.0   
..        ...   ...    ...   ...    ...    ...   ...     ...   ...    ...   
106   0.17120   0.0   8.56   0.0  0.520  5.836  91.9  2.2110   5.0  384.0   
270   0.29916  20.0   6.96   0.0  0.464  5.856  42.1  4.4290   3.0  223.0   
348   0.01501  80.0   2.01   0.0  0.435  6.635  29.7  8.3440   4.0  280.0   
435  11.16040   0.0  18.10   0.0  0.740  6.629  94.6  2.1247  24.0  666.0   
102   0.22876   0.0   8.56   0.0  0.520  6.405  85.4  2.7147   5.0  384.0   

     PTRATIO       B  LSTAT  
5       18.7  394.12   5.21  
116     17.8  393.30  12.04  
45      17.9  396.90  10.21  
16      21.0  386.85   6.58  
468     20.2  368.74  18.13  
..       ...     ...    ...  
106     20.9  395.67  18.66  
270     18.6  388.65  13.00  
348     17.0  390.94   5.99  
435     20.2  109.85  23.27  
102     20.9   70.80  10.63  

[354 rows x 13 columns]

In [15]:
y_test

price
5     28.7
116   21.2
45    19.3
16    23.1
468   19.1
..     ...
106   19.5
270   21.1
348   24.5
435   13.4
102   18.6

[354 rows x 1 columns]

In [16]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [17]:
lr.fit(x_train, y_train)

LinearRegression()

In [18]:
y_pred = lr.predict(x_test)
y_pred

array([[25.28344517],
       [22.89456654],
       [22.4267451 ],
       [19.21173569],
       [16.88542799],
       [20.783433  ],
       [27.74865322],
       [10.12612875],
       [26.54024713],
       [19.0973043 ],
       [19.9648131 ],
       [24.37761265],
       [30.19629642],
       [20.55699613],
       [40.83173071],
       [19.25570753],
       [15.99734981],
       [20.75879391],
       [11.70335184],
       [22.67092154],
       [23.22418964],
       [32.02732394],
       [21.40898069],
       [20.6208889 ],
       [24.92054404],
       [30.33492173],
       [31.48387946],
       [29.07394767],
       [ 9.23809869],
       [ 8.16380979],
       [13.68738909],
       [16.56424573],
       [32.27952145],
       [31.73287476],
       [17.0843591 ],
       [ 4.68316886],
       [29.23136307],
       [20.1880743 ],
       [17.74117357],
       [26.30874891],
       [18.74508275],
       [17.78992553],
       [12.65232927],
       [31.79458785],
       [22.51591798],
       [20

In [19]:
check = pd.DataFrame({'y_test': y_test['price'],
                     'y_pred': y_pred.flatten()})

In [20]:
check

y_test     y_pred
5      28.7  25.283445
116    21.2  22.894567
45     19.3  22.426745
16     23.1  19.211736
468    19.1  16.885428
..      ...        ...
106    19.5  18.048172
270    21.1  23.292194
348    24.5  27.699544
435    13.4  13.898729
102    18.6  20.955964

[354 rows x 2 columns]

***Для оценки точности полученной модели используем r2***

In [21]:
from sklearn.metrics import r2_score

In [22]:
print (f'Точность модели на тестовых данных после обучения на тренировочных : {round(r2_score(y_test,y_pred),6)}')

Точность модели на тестовых данных после обучения на тренировочных : 0.704786


###  Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
- Сделайте агрумент n_estimators равным 1000,
- Сделайте max_depth равным 12
- random_state сделайте равным 42.

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,но:
-  в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

Сделайте предсказание на тестовых данных и посчитайте R2.

Сравните с результатом из предыдущего задания.

Напишите в комментариях к коду, какая модель в данном случае работает лучше


In [23]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

In [24]:
model = RandomForestRegressor(n_estimators=1000, max_depth = 12, random_state = 42)

In [25]:
model.fit(x_train, y_train.values[:,0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [26]:
y_pred_rfr = model.predict(x_test)

In [27]:
check_rfr = pd.DataFrame({'y_test': y_test['price'],
                     'y_pred_lr': y_pred.flatten(),
                     'y_pred_rfr': y_pred_rfr.flatten()})

In [28]:
check_rfr

y_test  y_pred_lr  y_pred_rfr
5      28.7  25.283445   24.577202
116    21.2  22.894567   19.911643
45     19.3  22.426745   20.254981
16     23.1  19.211736   22.012374
468    19.1  16.885428   16.152984
..      ...        ...         ...
106    19.5  18.048172   17.014191
270    21.1  23.292194   19.896762
348    24.5  27.699544   25.506752
435    13.4  13.898729   12.614113
102    18.6  20.955964   21.215731

[354 rows x 3 columns]

In [29]:
print (f'Точность модели на алгоритме RandomForestRegressor на тестовых данных: {round(r2_score(y_test,y_pred_rfr),6)}')
print (f'Точность модели на  алгоритме LinearRegression на тестовых данных : {round(r2_score(y_test,y_pred),6)}')

Точность модели на алгоритме RandomForestRegressor на тестовых данных: 0.839206
Точность модели на  алгоритме LinearRegression на тестовых данных : 0.704786


**ОТВЕТ:*** *На мой взгляд регрессионная модель,построенная и обученная на алгоритме "Случайного Леса" более точная, чем линейная.*

### *Задание 3
- Вызовите документацию для класса RandomForestRegressor,найдите информацию об атрибуте feature_importances_.
- С помощью этого атрибута найдите сумму всех показателей важности,установите, какие два признака показывают наибольшую важность.

In [30]:
model.feature_importances_

array([0.05173963, 0.00080822, 0.00478773, 0.00061146, 0.03582382,
       0.31913431, 0.02758268, 0.03479234, 0.00179088, 0.02096416,
       0.01834726, 0.00858976, 0.47502773])

In [31]:
feature_imp = pd.DataFrame({'features': features,
                            'features_importances_': model.feature_importances_}).sort_values(by = 'features_importances_', ascending = False)
feature_imp.head(2)

features  features_importances_
12    LSTAT               0.475028
5        RM               0.319134

ОТВЕТ: Два признака, показывающие наибольшую важность - LSTAT (% lower status of the population) и RM (average number of rooms per dwelling) со значениями важности 0.475028 и 0.319134 соответственно.

### *Задание 4
В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию
по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.

Для этого датасета мы будем решать задачу классификации - будем определять,какие из транзакций по кредитной карте являются
мошенническими. Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки),так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель.

Мы будем вычислять AUC, то есть площадь под кривой ROC.

- Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.
- Загрузите датасет creditcard.csv и создайте датафрейм df.
- С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована.
- Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков.

Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:
- pd.options.display.max_columns = 100.
- Просмотрите первые 10 строк датафрейма df.
- Создайте датафрейм X из датафрейма df, исключив столбец Class.
- Создайте объект Series под названием y из столбца Class.
- Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.

У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.

Для поиска по сетке параметров задайте такие параметры:
- parameters = [{'n_estimators': [10, 15], 'max_features': np.arange(3, 5),'max_depth': np.arange(4, 7)}]

Создайте модель GridSearchCV со следующими аргументами:
- estimator=RandomForestClassifier(random_state=100),
- param_grid=parameters,
- scoring='roc_auc',
- cv=3.

Обучите модель на тренировочном наборе данных (может занять несколько минут).

Просмотрите параметры лучшей модели с помощью атрибута best_params_.

Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.

- Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba.
- Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
- Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументовмассивы y_test и y_pred_proba.

In [32]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

In [33]:
DATASET_PATH = './creditcard.csv'
df = pd.read_csv(DATASET_PATH, sep=',')

In [34]:
type(df)

pandas.core.frame.DataFrame

In [35]:
df.value_counts(normalize=True)

Time      V1          V2          V3         V4         V5         V6         V7          V8          V9         V10        V11        V12        V13        V14        V15        V16        V17        V18        V19        V20        V21         V22        V23        V24        V25        V26        V27        V28        Amount  Class
163152.0  -1.203617    1.574009    2.889277   3.381404   1.538663   3.698747   0.560211   -0.150911    0.124136   4.220998   1.384569  -0.706897  -0.256274  -1.562583   1.692915  -0.787338  -0.226776  -0.412354   0.234322   1.385597  -0.366727    0.522223  -0.357329  -0.870174  -0.134166   0.327019  -0.042648  -0.855262  1.51    0        0.000063
          -1.196037    1.585949    2.883976   3.378471   1.511706   3.717077   0.585362   -0.156001    0.122648   4.217934   1.385525  -0.709405  -0.256168  -1.564352   1.693218  -0.785210  -0.228008  -0.412833   0.234834   1.375790  -0.370294    0.524395  -0.355170  -0.869790  -0.133198   0.327804  -0.035702  -0

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [37]:
pd.options.display.max_columns = 100 # чтобы можно было легче просматривать все столбцы датафрейма

In [38]:
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   
5  0.260314 -0.568671 -0.371407  1.341262  0.359894 -0.358091 -0.137134   
6  0.081213  0.464960 -0.099254 -1.416907 -0.153826 -0.751063  0.167372   
7 -3.807864  0.615375  1.249376 -0.619468  0.291474  1.757964 -1.323865   
8  0.851084 -0.392048 -0.410430 -0.705117 -0.110452 -0.286254  0.074355   
9  0.069539 -0.736727 -0.366846  1.017614  0.836390  1.006844 -0.443523   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   
5  0.517617  0.401726 -0.058133  0.068653 -0.033194  0.084968 -0.208254   
6  0.050144 -0.443587  0.002821 -0.611987 -0.045575 -0.219633 -0.167716   
7  0.686133 -0.076127 -1.222127 -0.358222  0.324505 -0.156742  1.943465   
8 -0.328783 -0.210077 -0.499768  0.118765  0.570328  0.052736 -0.073425   
9  0.150219  0.739453 -0.540980  0.476677  0.451773  0.203711 -0.246914   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
5 -0.559825 -0.026398 -0.371427 -0.232794  0.105915  0.253844  0.081080   
6 -0.270710 -0.154104 -0.780055  0.750137 -0.257237  0.034507  0.005168   
7 -1.015455  0.057504 -0.649709 -0.415267 -0.051634 -1.206921 -1.085339   
8 -0.268092 -0.204233  1.011592  0.373205 -0.384157  0.011747  0.142404   
9 -0.633753 -0.120794 -0.385050 -0.069733  0.094199  0.246219  0.083076   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0  
5    3.67      0  
6    4.99      0  
7   40.80      0  
8   93.20      0  
9    3.68      0

In [39]:
x = df.drop(['Class'], axis=1)
x

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9       V10       V11       V12  \
0       0.462388  0.239599  0.098698  0.363787  0.090794 -0.551600 -0.617801   
1      -0.082361 -0.078803  0.085102 -0.255425 -0.166974  1.612727  1.065235   
2       1.800499  0.791461  0.247676 -1.514654  0.207643  0.624501  0.066084   
3       1.247203  0.237609  0.377436 -1.387024 -0.054952 -0.226487  0.178228   
4       0.095921  0.592941 -0.270533  0.817739  0.753074 -0.822843  0.538196   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  4.356170 -1.593105  2.711941   
284803  1.058415  0.024330  0.294869  0.584800 -0.975926 -0.150189  0.915802   
284804  3.031260 -0.296827  0.708417  0.432454 -0.484782  0.411614  0.063119   
284805  0.623708 -0.686180  0.679145  0.392087 -0.399126 -1.933849 -0.962886   
284806 -0.649617  1.577006 -0.414650  0.486180 -0.915427 -1.040458 -0.031513   

             V13       V14       V15       V16       V17       V18       V19  \
0      -0.991390 -0.311169  1.468177 -0.470401  0.207971  0.025791  0.403993   
1       0.489095 -0.143772  0.635558  0.463917 -0.114805 -0.183361 -0.145783   
2       0.717293 -0.165946  2.345865 -2.890083  1.109969 -0.121359 -2.261857   
3       0.507757 -0.287924 -0.631418 -1.059647 -0.684093  1.965775 -1.232622   
4       1.345852 -1.119670  0.175121 -0.451449 -0.237033 -0.038195  0.803487   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -0.689256  4.626942 -0.924459  1.107641  1.991691  0.510632 -0.682920   
284803  1.214756 -0.675143  1.164931 -0.711757 -0.025693 -1.221179 -1.545556   
284804 -0.183699 -0.510602  1.329284  0.140716  0.313502  0.395652 -0.577252   
284805 -1.042082  0.449624  1.962563 -0.608577  0.509928  1.113981  2.897849   
284806 -0.188093 -0.084316  0.041333 -0.302620 -0.660377  0.167430 -0.256117   

             V20       V21       V22       V23       V24       V25       V26  \
0       0.251412 -0.018307  0.277838 -0.110474  0.066928  0.128539 -0.189115   
1      -0.069083 -0.225775 -0.638672  0.101288 -0.339846  0.167170  0.125895   
2       0.524980  0.247998  0.771679  0.909412 -0.689281 -0.327642 -0.139097   
3      -0.208038 -0.108300  0.005274 -0.190321 -1.175575  0.647376 -0.221929   
4       0.408542 -0.009431  0.798278 -0.137458  0.141267 -0.206010  0.502292   
...          ...       ...       ...       ...       ...       ...       ...   
284802  1.475829  0.213454  0.111864  1.014480 -0.509348  1.436807  0.250034   
284803  0.059616  0.214205  0.924384  0.012463 -1.016226 -0.606624 -0.395255   
284804  0.001396  0.232045  0.578229 -0.037501  0.640134  0.265745 -0.087371   
284805  0.127434  0.265245  0.800049 -0.163298  0.123205 -0.569159  0.546668   
284806  0.382948  0.261057  0.643078  0.376777  0.008797 -0.473649 -0.818267   

             V27       V28  Amount  
0       0.133558 -0.021053  149.62  
1      -0.008983  0.014724    2.69  
2      -0.055353 -0.059752  378.66  
3       0.062723  0.061458  123.50  
4       0.219422  0.215153   69.99  
...          ...       ...     ...  
28480

In [40]:
y =df['Class']
y

0         0
1         0
2         0
3         0
4         0
         ..
284802    0
284803    0
284804    0
284805    0
284806    0
Name: Class, Length: 284807, dtype: int64

In [41]:
type(y)

pandas.core.series.Series

In [42]:
x_test,x_train,y_test,y_train = train_test_split(x, y, test_size = 0.3, random_state = 100,stratify = y)

In [43]:
print(' x_test : {} \n x_train : {}\n y_test :{} \n y_train : {}'.format(x_test.shape,x_train.shape,y_test.shape,y_train.shape))

 x_test : (199364, 30) 
 x_train : (85443, 30)
 y_test :(199364,) 
 y_train : (85443,)


In [44]:
parameters = [{'n_estimators': [10, 15],
               'max_features': np.arange(3, 5),
               'max_depth': np.arange(4, 7)}]

In [45]:
clf = GridSearchCV(estimator=RandomForestClassifier(random_state=100),
                   param_grid=parameters,
                   scoring='roc_auc',
                   cv=3)

In [46]:
clf.fit(x_train,y_train)

GridSearchCV(cv=3, estimator=RandomForestClassifier(random_state=100),
             param_grid=[{'max_depth': array([4, 5, 6]),
                          'max_features': array([3, 4]),
                          'n_estimators': [10, 15]}],
             scoring='roc_auc')

In [47]:
best_params = clf.best_params_
best_params

{'max_depth': 6, 'max_features': 4, 'n_estimators': 10}

In [48]:
prob1 =clf.predict_proba(x_test)
prob1

array([[9.99850339e-01, 1.49660702e-04],
       [9.99850339e-01, 1.49660702e-04],
       [9.99850339e-01, 1.49660702e-04],
       ...,
       [9.99784897e-01, 2.15102923e-04],
       [9.99784897e-01, 2.15102923e-04],
       [9.99789798e-01, 2.10201595e-04]])

In [49]:
pd.DataFrame(prob1,columns = df['Class'].unique())

0         1
0       0.999850  0.000150
1       0.999850  0.000150
2       0.999850  0.000150
3       0.999798  0.000202
4       0.999762  0.000238
...          ...       ...
199359  0.999785  0.000215
199360  0.999753  0.000247
199361  0.999785  0.000215
199362  0.999785  0.000215
199363  0.999790  0.000210

[199364 rows x 2 columns]

In [50]:
clf1 = RandomForestClassifier(random_state=100,
                              n_estimators = best_params['n_estimators'],
                              max_depth = best_params['max_depth'],
                              max_features = best_params['max_features'])

In [51]:
clf1.fit(x_train,y_train)

RandomForestClassifier(max_depth=6, max_features=4, n_estimators=10,
                       random_state=100)

In [52]:
y_pred = clf1.predict(x_test)
y_pred 

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [53]:
y_pr =clf1.predict_proba(x_test)
y_pr

array([[9.99850339e-01, 1.49660702e-04],
       [9.99850339e-01, 1.49660702e-04],
       [9.99850339e-01, 1.49660702e-04],
       ...,
       [9.99784897e-01, 2.15102923e-04],
       [9.99784897e-01, 2.15102923e-04],
       [9.99789798e-01, 2.10201595e-04]])

In [54]:
df_proba = pd.DataFrame(y_pred_proba,columns = df['Class'].unique())
df_proba

NameError: name 'y_pred_proba' is not defined

In [ ]:
y_pred_proba = df_proba[1]
y_pred_proba

In [ ]:
from sklearn.metrics import roc_auc_score

In [ ]:
roc_test = roc_auc_score(y_test,y_pred_proba)
roc_prob = roc_auc_score(y,y_pred_proba)

Не понятно,как считать roc_score,